src/jit/lower.cpp

```diff
@@ -571,6 +579,240 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
 #endif // //_TARGET_64BIT_
 }
 
+// Decompose 64-bit storeIndir tree into multiple 32-bit trees.
+#if !defined(_TARGET_64BIT_)
+void Lowering::DecomposeStoreInd(GenTree* tree)
+{
+    assert(tree->gtOp.gtOp2->OperGet() == GT_LONG);
+
+    GenTreeStmt* currStmt = comp->compCurStmt->AsStmt();
+    bool isEmbeddedStmt = !currStmt->gtStmtIsTopLevel();
+
+    // Example input trees (a nested embedded statement case)
+    //
+    //   <linkBegin Node>
+    //   *  stmtExpr  void  (top level) (IL   ???...  ???)
+    //   |  /--*  argPlace  ref    $280
+    //   |  +--*  argPlace  int    $4a
+    //   |  |  {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //   |  |  {  |     /--*  lclVar    ref    V11 tmp9         u:3 $21c
+    //   |  |  {  |     +--*  const     int    4 $44
+    //   |  |  {  |  /--*  +         byref  $2c8
+    //   |  |  {  |  |     {  *  stmtExpr  void  (embedded) (IL   ???...  ???)
+    //   |  |  {  |  |     {  |  /--*  lclFld    long   V01 arg1         u:2[+8] Fseq[i] $380
+    //   |  |  {  |  |     {  \--*  st.lclVar long  (P) V21 cse8
+    //   |  |  {  |  |     {  \--*    int    V21.hi (offs=0x00) -> V22 rat0    
+    //   |  |  {  |  |     {  \--*    int    V21.hi (offs=0x04) -> V23 rat1    
+    //   |  |  {  |  |  /--*  lclVar    int    V22 rat0          $380
+    //   |  |  {  |  |  +--*  lclVar    int    V23 rat1
+    //   |  |  {  |  +--*  gt_long   long
+    //   |  |  {  \--*  storeIndir long
+    //   |  +--*  lclVar    ref    V11 tmp9         u:3 (last use) $21c
+    //   |  +--*  lclVar    ref    V02 tmp0         u:3 $280
+    //   |  +--*  const     int    8 $4a
+    //   \--*  call help void   HELPER.CORINFO_HELP_ARRADDR_ST $205
+    //  <linkEndNode>
+
+    GenTree* linkBegin = comp->fgGetFirstNode(tree)->gtPrev;
+    GenTree* linkEnd = tree->gtNext;
+    GenTree* gtLong = tree->gtOp.gtOp2;
+
+    // Save address to a temp. It is used in storeIndLow and storeIndHigh trees.
+    GenTreeStmt* addrStmt = comp->fgInsertEmbeddedFormTemp(&tree->gtOp.gtOp1);
```

Comment:
>  I think that tree->gtOp.gtOp1 is already decomposed 

Yes, that tree is decomposed and it is moved to a new statement before the current statement so it won't be decomposed again. That's fine but the tree may already contain embedded statements and those too move to the new statement. Those embedded statements were supposed to be decomposed after the current statement but since they moved before it they'll never be decomposed.

This happened to me in lowering, see https://github.com/dotnet/coreclr/pull/1241#issuecomment-219940032. See `LowerArrElem`, those embedded statements are manually lowered to workaround the issue.

Since decomposition works exactly like lowering I expect this code to have the same issue. Unfortunately I've yet to produce a test case, I tried in https://github.com/dotnet/coreclr/pull/4986#issuecomment-220518856 but run into a different bug and anyway it's obvious that the test isn't good because the existing embedded tree doesn't contain long operations. But let's look at the tree:
```
*  stmtExpr  void  (top level) (IL 0x000...0x00B)
|           /--*  lclVar    int    V01 arg1         u:2 $c0
|           +--*  const     int    4 $43
|        /--*  +         int    $140
|        |  {  *  stmtExpr  void  (embedded) (IL 0x000...  ???)
|        |  {  |        /--*  lclVar    ref    V00 this         u:2 $80
|        |  {  |        +--*  const     int    4 field offset Fseq[i] $43
|        |  {  |     /--*  +         byref  $180
|        |  {  |  /--*  indir     int    <l:$142, c:$200>
|        |  {  \--*  st.lclVar int    V03 cse0
|        +--*  lclVar    int    V03 cse0          <l:$142, c:$200>
|     /--*  -         int    <l:$144, c:$143>
|  /--*  cast_ovfl long <- ulong <- int <l:$243, c:$242>
\--*  st.lclVar long   V02 loc0         d:3
```
In this example I need to make a temp out of the cast operand. Consider what would happen if the existing embedded statement contained long operations (it doesn't in the example).